contentcuration/contentcuration/utils/nodes.py

```diff
@@ -275,3 +286,189 @@ def generate_diff(updated_id, original_id):
+            # notice use of special `HGET`
+            # See: https://redis.io/commands/hget
+            return self.redis_client.hget(self.hash_key, key)
+        return self.cache.get("{}:{}".format(self.hash_key, key))
+
+    @redis_retry
+    def cache_set(self, key, val):
+        if self.redis_client is not None:
+            # notice use of special `HSET`
+            # See: https://redis.io/commands/hset
+            return self.redis_client.hset(self.hash_key, key, val)
+        return self.cache.set("{}:{}".format(self.hash_key, key), val)
+
+    def get_size(self):
+        size = self.cache_get(self.size_key)
+        return int(size) if size else size
+
+    def get_modified(self):
+        modified = self.cache_get(self.modified_key)
+        return isoparse(modified) if modified else modified
+
+    def set_size(self, size):
+        return self.cache_set(self.size_key, size)
+
+    def set_modified(self, modified):
+        return self.cache_set(self.modified_key, modified.isoformat() if isinstance(modified, datetime) else modified)
+
+
+class ResourceSizeHelper:
+    """
+    Helper class for calculating resource size
+    """
+    def __init__(self, node):
+        """
+        :param node: The contentnode with which to determine resource size
+        :type node: ContentNode
+        """
+        self.node = node
+
+    @property
+    def queryset(self):
+        """
+        :rtype: QuerySet
+        """
+        qs = self.node.get_descendants(include_self=True)
+
+        if self.node.is_root_node():
+            # since root node, remove unneeded filtering
+            qs = ContentNode.objects.filter(tree_id=self.node.tree_id)
+        # else if it's a leaf node, simplification handled by `get_descendants`
+
+        return File.objects.filter(contentnode__in=qs.filter(complete=True))
+
+    def get_size(self):
+        """
+        Calculates the size of the resource and it's descendants
+
+        SQL:
+            SELECT SUM(file_size)
+            FROM (
+                SELECT DISTINCT
+                    checksum,
+                    file_size
+                FROM contentcuration_file
+                WHERE contentnode_id IN ( SELECT id FROM contentcuration_contentnode WHERE ... )
+            ) subquery
+            ;
+
+        :return: An integer representing the resource size
+        """
+        sizes = self.queryset.values("checksum").distinct().aggregate(resource_size=Sum("file_size"))
+        return sizes['resource_size']
+
+    def modified_since(self, compare_datetime):
+        """
+        Determines if resources have been modified since ${compare_datetime}
+
+        SQL:
+            SELECT BOOL_OR(modified > ${compare_datetime})
+            FROM (
+                SELECT
+                    modified_at
+                FROM contentcuration_file
+                WHERE contentnode_id IN ( SELECT id FROM contentcuration_contentnode WHERE ... )
+            ) subquery
+            ;
```

Comment:
The use of the `BOOL_OR` aggregate with the index ordered descending by `modified` nulls last, this should hopefully be a quick query